src/core/lib/surface/byte_buffer_reader.cc

```diff
@@ -91,6 +91,23 @@ void grpc_byte_buffer_reader_destroy(grpc_byte_buffer_reader* reader) {
   }
 }
 
+int grpc_byte_buffer_reader_peek(grpc_byte_buffer_reader* reader,
+                                 grpc_slice** slice) {
+  switch (reader->buffer_in->type) {
```

Comment:
Two requests for tests?

1. Can you test this new surface API in `test/core/surface/byte_buffer_reader_test.cc` so that we have a documentation of its use and are confident of it TSAN-wise etc?
1. Can you add something to `test/cpp/microbenchmarks/bm_byte_buffer.cc` (or one of the other microbenchmarks, if more appropriate) to validate that this has a performance benefit on its own? That way bm_diff will also make sure that it doesn't get regressed. Even if it's just one test that's a loop of byte_buffer_reader_next and another one that's a loop of byte_buffer_reader_peek , it would be helpful.

Reply:
Yes, thanks for asking for tests. Completely forgot about those.

I added the tests and benchmarks. I will probably add a benchmark for protobuf in another patch.